src/client/Client.cc

```diff
@@ -8326,6 +8615,72 @@ int Client::opendir(const char *relpath, dir_result_t **dirpp, const UserPerm& p
   return r;
 }
 
+int Client::_fopendir(int dirfd, dir_result_t **dirpp, const UserPerm &perms) {
+  RWRef_t mref_reader(mount_state, CLIENT_MOUNTING);
+  if (!mref_reader.is_state_satisfied()) {
+    return -CEPHFS_ENOTCONN;
+  }
+
+  tout(cct) << __func__ << std::endl;
+  tout(cct) << dirfd << std::endl;
+
+  InodeRef dirinode;
+  int r = get_fd_inode(dirfd, &dirinode);
+  if (r < 0) {
+    return r;
+  }
+
+  if (cct->_conf->client_permissions) {
+    r = may_open(dirinode.get(), O_RDONLY, perms);
+    if (r < 0) {
+      return r;
+    }
+  }
+  r = _opendir(dirinode.get(), dirpp, perms);
+  /* if ENOTDIR, dirpp will be an uninitialized point and it's very dangerous to access its value */
+  if (r != -CEPHFS_ENOTDIR) {
+      tout(cct) << (uintptr_t)*dirpp << std::endl;
+  }
+  return r;
+}
+
+int Client::fopendir(int dirfd, dir_result_t **dirpp, const UserPerm &perms) {
+  std::scoped_lock locker(client_lock);
+  return _fopendir(dirfd, dirpp, perms);
+}
+
+int Client::opendirat(int dirfd, const char *relpath, int flags, dir_result_t **dirpp, const UserPerm &perms) {
```

Comment:
`opendirat` is non-standard, why not have the caller use `::openat`?

Reply:
basically, it just removed this call (callers can use openat + fdopendir)